[FMV][AArch64] Unify ls64, ls64_v and ls64_accdata.

labrinea opened the pull request:
Originally I tried spliting these features in the compiler with https://github.com/llvm/llvm-project/pull/101712, but we decided to lump those features in the ACLE specification (see https://github.com/ARM-software/acle/pull/346). Since there are no hardware implementations out there which implement ls64 without ls64_v or ls64_accdata, this shouldn't be a regression for feature detection.

labrinea committed:
[FMV][AArch64] Unify ls64, ls64_v and ls64_accdata.

Originally I tried spliting these features in the compiler with
https://github.com/llvm/llvm-project/pull/101712, but I realized
that there's no way to preserve backwards compatibility, therefore
we decided to lump those features in the ACLE specification too
https://github.com/ARM-software/acle/pull/346. Since there are no
hardware implementations out there which implement ls64 without
ls64_v or ls64_accdata, this shouldn't be a regression for feature
detection.

andrewcarlotti commented on compiler-rt/lib/builtins/cpu_model/AArch64CPUFeatures.inc:
```diff
@@ -73,8 +73,6 @@ enum CPUFeatures {
   FEAT_SSBS2,
   FEAT_BTI,
-  FEAT_LS64,
-  FEAT_LS64_V,
```
I think we might need to keep the enum values stable, although I'm not sure whether this could actually cause any issues yet.
(c.f. https://github.com/llvm/llvm-project/pull/88965#pullrequestreview-2004684899, which didn't reach any conclusion)

Wilco1 replied:
We **absolutely** must keep the values the same since they are now ABI and used in the ifunc resolvers in the object files:

```
        bl      __init_cpu_features_resolver
        adrp    x0, __aarch64_cpu_features
        ldr     x0, [x0, #:lo12:__aarch64_cpu_features]
        tbnz    x0, 36, .L8
```

The 36 is from the enum, now imagine a new compiler release changing it to 35 or a different compiler using 40...

labrinea replied:
Thanks for flagging this up. I am puzzled though since I was under the impression that FMV is still not production ready. When did the enum become part of the ABI? I guess a workaround is to keep a reserved position in the enum when removing a feature.

jroelofs replied:
https://github.com/llvm/llvm-project/pull/88965#discussion_r1567997481

jroelofs replied:
https://github.com/llvm/llvm-project/pull/88965#discussion_r1567997481

> Did any implementation ship __aarch64_cpu_features in a way that would make this an ABI break? We (Apple) talked about putting these in libSystem, and now I am glad we did not. It's okay to change/reorder them if it is a private implementation detail between the compiler and compiler-rt that is shipped with it, but as soon as they're part of an ABI, this list must become append only.

https://github.com/llvm/llvm-project/pull/88965#discussion_r1568423939

> For the time being __aarch64_cpu_features is not part of the ABI, but that may change in the future.

https://github.com/llvm/llvm-project/pull/88965#discussion_r1570826110

> For the ifunc implementation to work reliably, then the relevant parts of compiler-rt (or libgcc for the GCC implementation) should be statically linked. Otherwise there's a possibility that the symbols aren't actually loaded yet when the ifunc resolver is run. So there should be no exposed ABI in the context of FMV.

> As for other uses, we haven't agreed to freeze this as part of the ABI yet, but we are giving this possibility some consideration in designing the FMV resolver.

Wilco1 replied:
It became a defacto ABI when the first implementation became available. We could change the names of the global/init function to explicitly break the ABI with old objects at linktime to avoid silent ABI breaks that lead to incorrect execution at runtime.

jroelofs replied:
We could do that automatically every release by putting the version number in its name.

Wilco1 replied:
Yes, that's a way to keep backwards compatibility with older ABIs. However we can't break this ABI every compiler release (except perhaps this one time).

tmatheson-arm replied:
@Wilco1 could you explain the scenario you are worried about in more detail? What you posted above appears to be a snippet of a generated resolver, suggesting that the problematic scenario is:
- You have an old object file which contains a resolver.
- The resolver refers to bit 36 and has references to `__init_cpu_features_resolver` (defined in `compiler-rt`)
- You link against a new `compiler-rt`, which has changed the meaning of bit 36.

I don't think step 3 is allowed. The compiler and compiler-rt need to remain in sync, afaik?

There are no references to specific bit numbers outside of the generated resolvers and `__init_cpu_features_resolver`.

Wilco1 replied:
ABI compatibility means you can use different compilers to build object files, link them and they work correctly. If object A uses bit 36, object B uses bit 35 and the runtime uses bit 34, how is it ever going to work?

So using an enum in the implementation is really a bug. The FEAT_xxx bits are defined as constants for a good reason. And yes, these are ABI too!

tmatheson-arm replied:
>  If object A uses bit 36, object B uses bit 35 and the runtime uses bit 34, how is it ever going to work?

Using them for what? Where are these bits being passed across an object file boundary?

Wilco1 replied:
They are stored in the global `__aarch64_cpu_features` which is set by the init code and used by every FMV ifunc resolver.

labrinea commented:
This may sound like a bad idea in terms of future maintenance, but how about we guarded the enum entries I am removing with
`#if __FUNCTION_MULTI_VERSIONING_SUPPORT_LEVEL < __ARM_ACLE_VERSION(2024, 3, 0)` ?

labrinea committed:
Merge branch 'main' into fmv-unify-ls64

labrinea committed:
Changes from last revision:

Bring back FEAT_LS64 and FEAT_LS64_V as reserved bits to preserve
ABI compatibility.

labrinea committed:
Merge branch 'main' into fmv-unify-ls64

Wilco1 reviewed (approved):
Looks good to me

labrinea commented:
Just letting you know that I have reopened https://github.com/llvm/llvm-project/pull/101712 as a draft just to demonstrate that backwards compatibility can be preserved if we split the features (previously I thought it wasn't possible, but I found a way). However I couldn't convince people to revert the ACLE patch https://github.com/ARM-software/acle/pull/346 therefore I will be merging this patch by the end of today if there are no objections. Many thanks to everyone involved in the reviews!

labrinea committed:
Merge branch 'main' into fmv-unify-ls64
